zcash_primitives/src/note_encryption.rs

```diff
@@ -398,7 +398,8 @@ fn parse_note_plaintext_without_memo<P: consensus::Parameters>(
     }
 
     if let Some(derived_esk) = note.derive_esk() {
-        if (note.g_d * derived_esk) != *epk {
+        // This enforces that epk is a jubjub::SubgroupPoint.
+        if (note.g_d * derived_esk).to_bytes() != epk.to_bytes() {
```

Comment:
⚠️ This is an unintended consensus change!

The encoding of Jubjub points is not quite canonical: there are two possible representations of the identity point (0, 1) and of the point (0, -1). The latter is not relevant here because `note.g_d * derived_esk` is necessarily a `jubjub::SubgroupPoint`, and so cannot be (0, -1). However, it is possible for `derived_esk` to be 0. (After the end of the ZIP 212 grace period this would require finding a BLAKE2b preimage of a 512-bit encoding of zero (mod r<sub>J</sub>), but Canopy is not active on mainnet yet.) And so if `ephemeralKey` is the non-canonical encoding of (0, 1), this condition will be false for the new code and true for the old code in that case.

Reply:
The behaviour after this PR matches the behaviour in `zcashd`, because [this check happens on the C++ side](https://github.com/zcash/zcash/blob/6f9b281ef7a5bc28a99d56e97106489b6314fd6b/src/zcash/Note.cpp#L297-L306), and the result of `note.g_d * derived_esk` is necessarily encoded via `to_bytes` in order to [pass it back across the FFI](https://github.com/zcash/zcash/blob/6f9b281ef7a5bc28a99d56e97106489b6314fd6b/src/rust/src/rustzcash.rs#L484-L487). The only difference is that here, we parse the encoded `epk` and then re-serialize it, but `jubjub::ExtendedPoint::from_bytes` rejects non-canonical encodings, so it isn't possible for `jubjub::ExtendedPoint::to_bytes` to return a different encoding than was parsed.

Reply:
You're right. I will fix the spec to match the implementation.